samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,284 @@
+                    resource_dict.get("Properties", {}).get(template_field),
+                    stack_resource_detail.get("StackResourceDetail", {}).get("PhysicalResourceId", ""),
+                ):
+                    return False
+
+        return True
+
+    def _remove_unnecessary_fields(self, template_dict: Dict, linked_resources: List[str] = []) -> List[str]:
+        """
+        Fields skipped during template comparison because sync --code can handle the difference:
+        * CodeUri or ImageUri property of AWS::Serverless::Function
+        * ImageUri, S3Bucket, S3Key, S3ObjectVersion fields in Code property of AWS::Lambda::Function
+        * ContentUri property of AWS::Serverless::LayerVersion
+        * Content property of AWS::Lambda::LayerVersion
+        * DefinitionBody property of AWS::Serverless::Api
+        * BodyS3Location property of AWS::ApiGateway::RestApi
+        * DefinitionUri property of AWS::Serverless::HttpApi
+        * BodyS3Location property of AWS::ApiGatewayV2::Api
+        * DefinitionUri property of AWS::Serverless::StateMachine
+        * DefinitionS3Location property of AWS::StepFunctions::StateMachine
+        * SamResourceId in Metadata property of all resources
+
+        Parameters
+        ----------
+        template_dict: Dict
+            The unprocessed template dictionary
+        linked_resources: List[str]
+            The corresponding resources in the other template that got processed
+
+        Returns
+        -------
+        List[str]
+            The list of resource IDs that got changed during sanitization
+        """
+
+        resources = template_dict.get("Resources", {})
+        processed_resources: Set[str] = set()
+
+        for resource_logical_id in resources:
+            resource_dict = resources.get(resource_logical_id, {})
+            resource_type = resource_dict.get("Type")
+
+            if resource_type in SYNCABLE_RESOURCES:
+                processed_resource = self._remove_resource_field(
+                    resource_logical_id,
+                    resource_type,
+                    resource_dict,
+                    linked_resources,
+                )
+
+                if processed_resource:
+                    LOG.debug("Sanitized %s resource %s", resource_type, resource_logical_id)
+                    processed_resources.add(processed_resource)
+
+            # Remove SamResourceId metadata since this metadata does not affect any cloud behaviour
+            resource_dict.get("Metadata", {}).pop("SamResourceId", None)
+            if not resource_dict.get("Metadata"):
+                resource_dict.pop("Metadata", None)
+            LOG.debug("Sanitizing the Metadata for resource %s", resource_logical_id)
+
+        return sorted(list(processed_resources))
+
+    def _remove_resource_field(
+        self,
+        resource_logical_id: str,
+        resource_type: str,
+        resource_dict: Dict,
+        linked_resources: List[str] = [],
```

Comment:
To improve the performance, this `linked_resources` can be a `Set[str]` instead of `List[str]`